### PR TITLE
Enable ViTReciproCAM for VisionTransformer backbone (cls task)

### PR DESCRIPTION
**Pull request by negvet**
### Summary

Enable ViTReciproCAM to support cls models with `VisionTransformer` backbone, e.g. deit.

Can be merged, but cannot be actually used in Geti until https://github.com/openvinotoolkit/training_extensions/issues/2098 is resoled (featute vector and saliency maps are coupled). Can be used right away only for torch runtime.

Validated in OTX:
- Torch task (infer, explain APIs).
- Export to OV IR.
- OV task (infer (async, sync), explain APIs).

Torch task examples:
image

OV task examples:
image

Multilabel:
image

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

**Review by jaegukhyun — DISMISSED** [review]
LGTM

File: src/otx/algorithms/common/adapters/mmcv/hooks/recording_forward_hook.py [hunk 1]
```diff
@@ -238,3 +239,122 @@ def _get_mosaic_feature_map(self, feature_map: torch.Tensor, c: int, h: int, w:
                     mosaic_feature_map_mask[k, :, i, j] = torch.ones(c).to(feature_map.device)
             mosaic_feature_map = feature_map_repeated * mosaic_feature_map_mask
         return mosaic_feature_map
+
+
+class ViTReciproCAMHook(BaseRecordingForwardHook):
```
Comment (eunwoosh) [inline src/otx/algorithms/common/adapters/mmcv/hooks/recording_forward_hook.py:244]: Could you add comments to explain each argument of init function?

Comment (eunwoosh) [inline src/otx/algorithms/common/adapters/mmcv/hooks/recording_forward_hook.py:244]: Could you add unit test for this class?

File: src/otx/algorithms/classification/adapters/mmcls/models/classifiers/custom_image_classifier.py [hunk 2]
```diff
@@ -300,6 +304,57 @@ def extract_feat(self, img):
         ReciproCAMHook,
     )
 
+    def _extract_vit_feat(model, x):
```
Comment (eunwoosh) [inline src/otx/algorithms/classification/adapters/mmcls/models/classifiers/custom_image_classifier.py:307]: Could you update or add unit test for this?

**Review by eunwoosh — COMMENTED** [review]
Thanks for your effort :) I left some comments. please take a look.

File: src/otx/algorithms/classification/adapters/mmcls/models/classifiers/custom_image_classifier.py [hunk 3]
```diff
@@ -300,6 +304,57 @@ def extract_feat(self, img):
...
+                    patch_token = patch_token.permute(0, 3, 1, 2)
+                    cls_token = None
+                if model.backbone.output_cls_token:
+                    feat = [patch_token, cls_token]
+                else:
+                    feat = patch_token
+        backbone_feat = feat
+        if model.with_neck:
+            feat = model.neck(feat)
+        return feat, backbone_feat, layernorm_feat
```
Comment (sovrasov) [inline src/otx/algorithms/classification/adapters/mmcls/models/classifiers/custom_image_classifier.py]: Could you rename feat->logits if this tensor represents logits?

Comment (sovrasov) [inline src/otx/algorithms/classification/adapters/mmcls/models/classifiers/custom_image_classifier.py]: Also, I propose returning cls token as a feature vector

File: src/otx/algorithms/common/adapters/mmcv/hooks/recording_forward_hook.py (hunk 1 above)
Comment (negvet) [reply src/otx/algorithms/common/adapters/mmcv/hooks/recording_forward_hook.py:244]: Sure, added

Comment (negvet) [reply src/otx/algorithms/common/adapters/mmcv/hooks/recording_forward_hook.py:244]: It is there, check out https://github.com/openvinotoolkit/training_extensions/blob/ceb0ffa8e7c436406578ae2d5238ec67a72e6626/tests/unit/algorithms/classification/test_xai_classification_validity.py#L51

Does it look good?

File: src/otx/algorithms/classification/adapters/mmcls/models/classifiers/custom_image_classifier.py (hunk 2 above)
Comment (negvet) [reply src/otx/algorithms/classification/adapters/mmcls/models/classifiers/custom_image_classifier.py:307]: Thanks, added

File: src/otx/algorithms/classification/adapters/mmcls/models/classifiers/custom_image_classifier.py (hunk 3 above)
Comment (negvet) [reply src/otx/algorithms/classification/adapters/mmcls/models/classifiers/custom_image_classifier.py]: Here, feat is not a logits yet. Logits are obtained at the next step `logit = self.head.simple_test(feat)`

Comment (negvet) [reply src/otx/algorithms/classification/adapters/mmcls/models/classifiers/custom_image_classifier.py]: Ok, updated

File: src/otx/algorithms/classification/adapters/mmcls/models/classifiers/custom_image_classifier.py [hunk 4]
```diff
@@ -320,12 +375,22 @@ def sam_image_classifier__extract_feat(ctx, self, img):  # pylint: disable=unuse
     )
     def sam_image_classifier__simple_test(ctx, self, img, img_metas):  # pylint: disable=unused-argument
         """Simple test function used for inference for SAMClassifier with mmdeploy."""
-        feat, backbone_feat = self.extract_feat(img)
+        vit_backbone = isinstance(self.backbone, VisionTransformer)
+        if vit_backbone:
+            feat, backbone_feat, layernorm_feat = _extract_vit_feat(self, img)
```
Comment (sovrasov) [inline src/otx/algorithms/classification/adapters/mmcls/models/classifiers/custom_image_classifier.py]: Why do wee need to return backbone_feat in this case?

Comment (negvet) [reply src/otx/algorithms/classification/adapters/mmcls/models/classifiers/custom_image_classifier.py]: We do not need. Thanks for noticing, fixed

File: src/otx/algorithms/common/adapters/mmcv/hooks/recording_forward_hook.py (hunk 1 above)
Comment (eunwoosh) [reply src/otx/algorithms/common/adapters/mmcv/hooks/recording_forward_hook.py:244]: Yes, it checks the function in some degree, but this function has a quite complex code. so, how about to add a little more unit test?

**Review by jaegukhyun — DISMISSED** [review]

File: src/otx/algorithms/common/adapters/mmcv/hooks/recording_forward_hook.py (hunk 1 above)
Comment (negvet) [reply src/otx/algorithms/common/adapters/mmcv/hooks/recording_forward_hook.py:244]: I added some unit tests, please take a look

**Review by sovrasov — APPROVED** [review]

**Review by sungmanc — APPROVED** [review]

**Review by eunwoosh — APPROVED** [review]
